File: packages/game-schema/src/__tests__/schema.test.ts
import { describe, it } from 'node:test'
import assert from 'node:assert/strict'
import {
  validateGameSchema,
  safeValidateGameSchema,
  assertSimulationReady,
  SCHEMA_VERSION,
} from '../index.ts'

const validSchema = {
  schemaVersion: SCHEMA_VERSION,
  provider: 'test-provider',
  gameId: 'test-game-001',
  gameName: 'Test Slot',
  gameType: 'video-slot' as const,
  currencyMode: 'credits' as const,
  bet: { defaultBet: 1, lines: 10, coinValue: 1 },
  reels: [
    ['A', 'B', 'C', 'WILD', 'A', 'B'],
    ['A', 'B', 'C', 'A', 'B', 'C'],
    ['A', 'B', 'C', 'SCATTER', 'A', 'B'],
    ['A', 'B', 'C', 'A', 'B', 'C'],
    ['A', 'B', 'C', 'A', 'WILD', 'B'],
  ],
  paylines: [
    [1, 1, 1, 1, 1],
    [0, 0, 0, 0, 0],
    [2, 2, 2, 2, 2],
  ],
  symbols: [
    { id: 'A', name: 'SymbolA', isWild: false, isScatter: false },
    { id: 'B', name: 'SymbolB', isWild: false, isScatter: false },
    { id: 'C', name: 'SymbolC', isWild: false, isScatter: false },
    { id: 'WILD', name: 'Wild', isWild: true, isScatter: false },
    { id: 'SCATTER', name: 'Scatter', isWild: false, isScatter: true },
  ],
  paytable: {
    A: { '3': 5, '4': 20, '5': 100 },
    B: { '3': 3, '4': 10, '5': 50 },
    C: { '3': 2, '4': 8, '5': 30 },
  },
  wild: { symbolId: 'WILD', substitutesFor: [] },
  scatter: { symbolId: 'SCATTER', triggerCount: 3, awardType: 'freeSpins' as const },
  freeSpins: { count: 10, multiplier: 2, retrigger: false },
  sourceEvidence: [],
  warnings: [],
  assumptions: [],
}

describe('GameSchema validation', () => {
  it('accepts a valid schema', () => {
    const result = validateGameSchema(validSchema)
    assert.equal(result.gameId, 'test-game-001')
    assert.equal(result.schemaVersion, SCHEMA_VERSION)
  })

  it('rejects a schema with wrong version', () => {
    const result = safeValidateGameSchema({ ...validSchema, schemaVersion: '9.9.9' })
    assert.equal(result.success, false)
  })

  it('rejects a schema with empty reels', () => {
    const result = safeValidateGameSchema({ ...validSchema, reels: [] })
    assert.equal(result.success, false)
  })

  it('rejects a schema with empty paylines', () => {
    const result = safeValidateGameSchema({ ...validSchema, paylines: [] })
    assert.equal(result.success, false)
  })

  it('rejects a schema missing gameName', () => {
    const { gameName: _, ...rest } = validSchema
    const result = safeValidateGameSchema(rest)
    assert.equal(result.success, false)
  })

  it('defaults warnings and assumptions to empty arrays', () => {
    const { warnings: _w, assumptions: _a, ...rest } = validSchema
    const result = validateGameSchema(rest)
    assert.deepEqual(result.warnings, [])
    assert.deepEqual(result.assumptions, [])
  })
})

describe('assertSimulationReady', () => {
  it('returns no errors for valid schema', () => {
    const schema = validateGameSchema(validSchema)
    const errors = assertSimulationReady(schema)
    assert.deepEqual(errors, [])
  })

  it('catches payline length mismatch', () => {
    const bad = validateGameSchema({
      ...validSchema,
      paylines: [[1, 1, 1]], // only 3 entries for 5 reels
    })
    const errors = assertSimulationReady(bad)
    assert.ok(errors.some((e) => e.includes('paylines[0]')))
  })

  it('catches unknown symbol in reel strip', () => {
    const bad = validateGameSchema({
      ...validSchema,
      reels: [
        ['A', 'UNKNOWN_SYM'],
        ...validSchema.reels.slice(1),
      ],
    })
    const errors = assertSimulationReady(bad)
    assert.ok(errors.some((e) => e.includes('UNKNOWN_SYM')))
  })

  it('catches missing paytable entry for non-wild non-scatter symbol', () => {
    const { A: _a, ...paytableWithoutA } = validSchema.paytable
    const bad = validateGameSchema({ ...validSchema, paytable: paytableWithoutA })
    const errors = assertSimulationReady(bad)
    assert.ok(errors.some((e) => e.includes('"A"')))
  })
})

File: services/simulator/internal/schema/schema_test.go
package schema

import (
	"encoding/json"
	"testing"
)

func minimalSchema() GameSchema {
	return GameSchema{
		SchemaVersion: "0.1.0",
		GameID:        "test",
		GameName:      "Test",
		GameType:      "video-slot",
		Bet:           BetConfig{DefaultBet: 1, Lines: 1, CoinValue: 1},
		Reels:         [][]string{{"A", "B"}, {"A", "B"}, {"A", "B"}},
		Paylines:      [][]int{{0, 0, 0}},
		Symbols: []Symbol{
			{ID: "A", Name: "A"},
			{ID: "B", Name: "B"},
		},
		Paytable: map[string]map[string]float64{
			"A": {"3": 10},
			"B": {"3": 5},
		},
	}
}

func TestValidate_Minimal(t *testing.T) {
	s := minimalSchema()
	if err := s.Validate(); err != nil {
		t.Fatalf("minimal schema should be valid: %v", err)
	}
}

func TestValidate_EmptyReels(t *testing.T) {
	s := minimalSchema()
	s.Reels = nil
	if err := s.Validate(); err == nil {
		t.Fatal("expected error for empty reels")
	}
}

func TestValidate_PaylineLengthMismatch(t *testing.T) {
	s := minimalSchema()
	s.Paylines = [][]int{{0, 0}}
	if err := s.Validate(); err == nil {
		t.Fatal("expected error for payline length mismatch")
	}
}

func TestValidate_UnknownSymbolInReel(t *testing.T) {
	s := minimalSchema()
	s.Reels[0] = []string{"A", "ZZZ"}
	if err := s.Validate(); err == nil {
		t.Fatal("expected error for unknown symbol")
	}
}

func TestValidate_MissingPaytableEntry(t *testing.T) {
	s := minimalSchema()
	delete(s.Paytable, "B")
	if err := s.Validate(); err == nil {
		t.Fatal("expected error for missing paytable entry")
	}
}

func TestValidate_WildScatterSkipPaytable(t *testing.T) {
	s := minimalSchema()
	s.Symbols = append(s.Symbols, Symbol{ID: "W", Name: "Wild", IsWild: true})
	s.Symbols = append(s.Symbols, Symbol{ID: "S", Name: "Scatter", IsScatter: true})
	// Add to reels so reel-symbol check passes.
	s.Reels[0] = []string{"A", "W", "S"}
	if err := s.Validate(); err != nil {
		t.Fatalf("wild/scatter should not need paytable entry: %v", err)
	}
}

func TestSimConfig_Defaults(t *testing.T) {
	c := SimulationConfig{}
	if err := c.Validate(); err != nil {
		t.Fatalf("default config should validate: %v", err)
	}
	if c.SpinCount != 10_000_000 {
		t.Fatalf("default spinCount = %d", c.SpinCount)
	}
	if c.Rows != 3 {
		t.Fatalf("default rows = %d", c.Rows)
	}
}

func TestSimConfig_DisallowedSpinCount(t *testing.T) {
	c := SimulationConfig{SpinCount: 12345}
	if err := c.Validate(); err == nil {
		t.Fatal("expected error for disallowed spin count")
	}
}

func TestSimConfig_AllowedSpinCounts(t *testing.T) {
	allowed := []int64{1_000_000, 10_000_000, 100_000_000, 500_000_000, 1_000_000_000}
	for _, n := range allowed {
		c := SimulationConfig{SpinCount: n}
		if err := c.Validate(); err != nil {
			t.Fatalf("spinCount %d should be allowed: %v", n, err)
		}
	}
}

func TestTotalBet(t *testing.T) {
	s := minimalSchema()
	s.Bet = BetConfig{DefaultBet: 0.5, Lines: 20, CoinValue: 1}
	if got := s.TotalBet(); got != 10 {
		t.Fatalf("TotalBet = %v want 10", got)
	}
	s.Bet.CoinValue = 0
	if got := s.TotalBet(); got != 10 {
		t.Fatalf("TotalBet with zero coin = %v want 10", got)
	}
}

func TestRoundtripJSON(t *testing.T) {
	in := minimalSchema()
	in.Wild = &WildConfig{SymbolID: "W", SubstitutesFor: []string{"A"}, Multiplier: 1}
	in.Scatter = &ScatterConfig{SymbolID: "S", TriggerCount: 3, AwardType: "freeSpins"}
	in.FreeSpins = &FreeSpinsConfig{Count: 10, Multiplier: 2, Retrigger: true}

	b, err := json.Marshal(in)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var out GameSchema
	if err := json.Unmarshal(b, &out); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if out.Wild == nil || out.Wild.SymbolID != "W" {
		t.Fatal("wild lost in roundtrip")
	}
	if out.Scatter == nil || out.Scatter.TriggerCount != 3 {
		t.Fatal("scatter lost in roundtrip")
	}
	if out.FreeSpins == nil || out.FreeSpins.Count != 10 {
		t.Fatal("freeSpins lost in roundtrip")
	}
}

func TestValidate_WaysMechanicAllowsEmptyPaylines(t *testing.T) {
	s := GameSchema{
		SchemaVersion: "0.1.0",
		GameID:        "test",
		GameName:      "test",
		Mechanic:      "ways",
		Bet:           BetConfig{DefaultBet: 1, Lines: 1, CoinValue: 1},
		Reels:         [][]string{{"A"}, {"A"}, {"A"}},
		Paylines:      [][]int{},
		Symbols:       []Symbol{{ID: "A", Name: "A"}},
		Paytable:      map[string]map[string]float64{"A": {"8": 20}},
	}
	if err := s.Validate(); err != nil {
		t.Fatalf("ways mechanic should allow empty paylines, got: %v", err)
	}
}

func TestValidate_RandomScatterInjectSymbolMustExist(t *testing.T) {
	s := GameSchema{
		SchemaVersion: "0.1.0",
		GameID:        "test",
		GameName:      "test",
		Mechanic:      "ways",
		Bet:           BetConfig{DefaultBet: 1, Lines: 1, CoinValue: 1},
		Reels:         [][]string{{"A"}, {"A"}, {"A"}},
		Paylines:      [][]int{},
		Symbols:       []Symbol{{ID: "A", Name: "A"}},
		Paytable:      map[string]map[string]float64{"A": {"8": 20}},
		RandomScatterInject: &RandomScatterInjectConfig{
			SymbolID:    "MISSING",
			BaseWeights: []ScatterWeightEntry{{Count: 1, Weight: 10}},
		},
	}
	if err := s.Validate(); err == nil {
		t.Fatal("expected error for unknown scatter inject symbol")
	}
}
